[WIP] Post review

pkseverance opened the pull request:
Summary of changes: Reviews can now be posted for a given shelter id. Error handling and testing have yet to be implemented, as I need to read up on error handling in express. I would prefer to wait until we've setup a testing environment to write tests, since it seems like good practice to keep development and testing data  in separate databases. 

This endpoint is a work in progress. Please don't merge yet.

Tscasady commented on src/controllers/reviews_controller.ts:
```diff
@@ -0,0 +1,24 @@
+        staff: staff,
+      },
+    });
+    res.status(204).send();
```
I think we should send back the newly created object in the body of the json response.

pkseverance replied:
Agreed. Thanks!

Tscasady reviewed (changes requested):
A good start, just going to wait on tests, see the PR for #4.

pkseverance commented on src/controllers/reviews_controller.ts:
```diff
@@ -0,0 +1,24 @@
+export const reviewController = express.Router();
+
+const prisma = new PrismaClient();
+
```
Thanks! Will do.

pkseverance commented:
Post reviews should now be functional. I implemented basic error handling, did some refactoring, and implemented unit testing for review post requests. Feel free to merge now.

Tscasady commented on package.json:
```diff
@@ -7,7 +7,8 @@
     "dev": "ts-node-dev --respawn --pretty --transpile-only src/server.ts",
     "testLocal": "dotenv -e .env.test npm run test",
     "test": "mocha --file ./test/helper.ts --require ts-node/register test/**/*-test.ts",
-    "start": "tsc && node dist/server.js"
```
This start script ought not be removed, it is used in deployment. I assume this was just a merge conflict error.

Tscasady commented on test/requests/reviews/reviewRequest-test.ts:
```diff
@@ -0,0 +1,79 @@
+
+      expect(res.body.data.attributes).to.have.key('staff')
+      expect(res.body.data.attributes.staff).to.eq(9.8)
+
```
I think we ought to include a test that has some null values, demonstrating that a user does not have to rate all aspects of a shelter.

Tscasady commented on test/requests/reviews/reviewRequest-test.ts:
```diff
@@ -0,0 +1,79 @@
+      staff: 9.8 
+    }
+
+    chai
```
All of these tests are giving false positives given the way you have handled async execution. For example I added 
```expect(res.body.data.id).to.equal(12312322) ``` And it passed. In the words of Brian, never trust a passing test! Make sure you get it to fail first.

Tscasady commented on src/serializers/review.ts:
```diff
@@ -0,0 +1,18 @@
+      },
+    }
+  } else {
+    return { data: {} }
```
This implementation was a good placeholder for 'errors' but I think now we are in a position to return real 404 errors and validation errors instead of simply an empty object. See #20 and #21

Tscasady replied:
If you would like to make a separate issue and PR for handling errors that is fine with me.

pkseverance replied:
I'm going to work on error handling separately.

Tscasady reviewed (changes requested):
I left some notes for how to proceed with error handling / validation, an example was put up today in other PRs. I'm mostly concerned with testing, your current test file is giving false positives.

pkseverance commented:
Everything should work now.

Tscasady commented on package.json:
```diff
@@ -6,8 +6,11 @@
     "testLocal": "dotenv -e .env.test npm run test",
-    "test": "mocha --file ./test/helper.ts --require ts-node/register 'test/**/*-test.ts'",
-    "start": "tsc && node dist/server.js"
+    "test": "mocha --file ./test/helper.ts --require ts-node/register test/**/*-test.ts",
```
you've removed the ' quotes around the 'test/**/*-test.ts' and it is no longer properly running all tests. Your most recent github actions job only ran two tests.

Tscasady commented on test/requests/reviews/reviewRequest-test.ts:
```diff
@@ -0,0 +1,67 @@
+  zip: number
+  phoneNumber: string
+}
+
```
These type declarations are now available in shelter.types.ts and review.types.ts

Tscasady commented on test/requests/shelters/shelterRequest-test.ts:
```diff
@@ -69,6 +77,56 @@ describe('Post Shelter', () => {
+      phoneNumber: '2134568765',
+      websiteUrl: "www.fake.com"
+    }
+
```
This file has had some significant changes, and the version on main is correct. This file is not related to this PR and is out of date.

Tscasady reviewed (approved):
Looks like you got all the testing sorted, and the CI checks look correct. Make sure you add an issue/card to project board for error/sad path for POST review.
